include/deal.II/lac/affine_constraints.templates.h

```diff
@@ -4592,6 +4592,11 @@ AffineConstraints<number>::distribute_local_to_global(
       if (n_values > 0)
         global_matrix.add(row, n_values, cols.data(), vals.data(), false, true);
     }
+
+  Vector<typename MatrixType::value_type> dummy(0);
+
+  internal::AffineConstraints::set_matrix_diagonals(
+    global_rows, row_indices, local_matrix, *this, global_matrix, dummy, false);
```

Comment:
I did not read the whole context yet, but is this what we desire? The question is if we expect the present contribution to be added as a diagonal block or not. For a DG face integral, we might have assembled an off-diagonal block with `row_indices != col_indices` that should not anything to the diagonal, but we might also add something with `row_indices == col_indices`. The reason why https://github.com/dealii/dealii/blob/f6fb6e8f9c339922de56b003ca27c3c63da03df1/include/deal.II/lac/affine_constraints.h#L1499-L1507 has the optional parameter is precisely to let the user control this situation. At this point, I would intuitively say we need to add this optional parameter to all other `distribute_local_to_global` functions that have two indices vectors. Without change in the interface, we would at least have to adapt the documentation because it says: https://github.com/dealii/dealii/blob/f6fb6e8f9c339922de56b003ca27c3c63da03df1/include/deal.II/lac/affine_constraints.h#L1662-L1672

Reply:
I agree with @kronbichler that this does not look right.

Reply:
I have opened an alternative PR: https://github.com/dealii/dealii/pull/17149. 

The problem is that in https://github.com/dealii/dealii/pull/17029, we have generalized `MatrixFreeTools::compute_matrix()` so that it can also compute off-diagonal blocks. While doing so, I have switched from one version of `AffineConstraints::distribute_local_to_global()` and another version, without knowing that the diagonal entries are treated differently, resulting in zero entries on the diagonal, with the result that the resulting matrix is singular. Now different application codes are crashing because the setup of the matrix-bases preconditioners are failing.

Reply:
I think the other PR is the right choice at this point. The remaining question is whether we want to look over the interfaces in `distribute_local_to_global`, as we have accumulated quite a number of options there, and see if this is what we desire.